src/library/scala/collection/immutable/OrderedMap.scala

```diff
@@ -0,0 +1,426 @@
+    val (head, tail) = ordering.headTail
+    new OrderedMap(tail, mapping.remove(head), ordinal, orderedBy)
+  }
+
+  override def init: OrderedMap[K, V] = {
+    val (init, last) = ordering.initLast
+    new OrderedMap(init, mapping.remove(last), ordinal, orderedBy)
+  }
+
+  def get(key: K): Option[V] = mapping.get(key).map(value)
+
+  def iterator: Iterator[(K, V)] = new AbstractIterator[(K, V)] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): (K, V) = binding(iter.next())
+  }
+
+  override def keysIterator: Iterator[K] = new AbstractIterator[K] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): K = iter.next()
+  }
+
+  override def valuesIterator: Iterator[V] = new AbstractIterator[V] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): V = value(binding(iter.next()))
+  }
+
+  override def concat[V1 >: V](suffix: IterableOnce[(K, V1)]): OrderedMap[K, V1] = {
+    var result: OrderedMap[K, V1] = this
+    val iter = suffix.iterator
+    while (iter.hasNext) {
+      val (k, v) = iter.next()
+      result = result.updated(k, v)
+    }
+    result
+  }
+
+  override def map[K2, V2](f: ((K, V)) => (K2, V2)): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    val mng: Mapping[K2, V2] = mapping.map {
+      case (k, (o, v)) =>
+        val (k2, v2) = f((k, v))
+        ong = ong.incl(o, k2)
+        (k2, (o, v2))
+    }
+    new OrderedMap[K2, V2](ong, mng, ordinal, orderedBy)
+  }
+
+  override def flatMap[K2, V2](f: ((K, V)) => IterableOnce[(K2, V2)]): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    val mng: Mapping[K2, V2] = mapping.flatMap {
+      case (k, (o, v)) =>
+        f((k, v)).iterator.reduceLeftOption((_, e) => e).map {
+          case (k2, v2) =>
+            ong = ong.incl(o, k2)
+            (k2, (o, v2))
+        }
+    }
+    new OrderedMap[K2, V2](ong, mng, ordinal, orderedBy)
+  }
+
+  override def collect[K2, V2](pf: PartialFunction[(K, V), (K2, V2)]): OrderedMap[K2, V2] = {
+    var ong: Ordering[K2] = Ordering.empty
+    val mng: Mapping[K2, V2] = mapping.collect {
+      case (k, (o, v)) if pf.isDefinedAt((k, v)) =>
+        val (k2, v2) = pf((k, v))
+        ong = ong.incl(o, k2)
+        (k2, (o, v2))
+    }
+    new OrderedMap[K2, V2](ong, mng, ordinal, orderedBy)
+  }
+
+  @`inline` private[this] def value(p: (_, V)) = p._2
+  @`inline` private[this] def binding(k: K) = mapping(k).copy(_1 = k)
+}
+object OrderedMap extends MapFactory[OrderedMap] {
+  sealed trait OrderBy
+  final object OrderBy {
+    final case object Insertion extends OrderBy
+    final case object Modification extends OrderBy
+  }
+
+  val Empty = new OrderedMap[Nothing, Nothing](Ordering.empty, HashMap.empty, 0, OrderBy.Insertion)
```

Comment:
I would recommend starting `ordinal` at `Int.MinValue`, rather than `0`. If you start at `0` and have a map containing just below `Int.MaxValue` elements, adding to and removing from that map is going to cause a rebuilding of the map every few additions, which would be extremely expensive.

Alternatively, you could use a `Long` for `ordinal`; I don't know to what degree that would impact performance or allocation size.

Reply:
The reason that the ordinal counting starts with `0: Int`, jumps from `Int.MaxValue` to `Int.MinValue` and repopulates a fresh instance once it reaches `-1: Int` is that the `IntMap` used for the ordering uses a tree structure that is based on the bit sequence of the `Int` keys. This means that `0: Int` (`00000000000000000000000000000000`) will be the leftmost (first) possible leaf in the tree, `Int.MaxValue` (`01111111111111111111111111111111`) will be followed directly by `Int.MinValue` (`10000000000000000000000000000000`), and the rightmost (last) possible leaf is `-1: Int` (`11111111111111111111111111111111`).

```scala
scala> IntMap(1 -> "one", 10 -> "ten", Int.MinValue -> "min", -10 -> "-ten", 0 -> "zero", -1 -> "-one", Int.MaxValue -> "max")
res55: scala.collection.immutable.IntMap[String] = IntMap(0 -> zero, 1 -> one, 10 -> ten, 2147483647 -> max, -2147483648 -> min, -10 -> -ten, -1 -> -one)
```

Reply:
ah, I didn't realize that. Thanks for the explanation!